### PR TITLE
ota-requestor-app - initial set of changes of development branch

### PR DESCRIPTION
**Pull request by holbrookt**
#### Problem
Submitting a work-in-progress ota-requestor-app so that others may begin to pick up development.
 
NOTE - this will not build successfully right now. See `examples/ota-requestor-app/linux/README.md` for the current state of this application.

#### Change overview
Add new example app `ota-requestor-app`.

#### Testing
N/A

### CODE REVIEW

File: examples/ota-requestor-app/linux/.gn [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# Copyright (c) 2020 Project CHIP Authors
```
Comment (BroderickCarlin) [inline examples/ota-requestor-app/linux/.gn:1]:
```suggestion
# Copyright (c) 2021 Project CHIP Authors
```

File: examples/ota-requestor-app/linux/BUILD.gn [hunk 2]
```diff
@@ -0,0 +1,39 @@
+# Copyright (c) 2020 Project CHIP Authors
```
Comment (BroderickCarlin) [inline examples/ota-requestor-app/linux/BUILD.gn:1]:
```suggestion
# Copyright (c) 2021 Project CHIP Authors
```

Comment (BroderickCarlin) [reply examples/ota-requestor-app/linux/BUILD.gn:1]: Looks like most of the copyrights need to be updated; I won't go through and comment on them all but they should be audited

Comment (holbrookt) [reply examples/ota-requestor-app/linux/BUILD.gn]: Done. Keep in mind that `build_overrides` files are symlinks

**Review by selissia — APPROVED** [review]

File: examples/ota-requestor-app/linux/PersistentStorage.cpp [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    std::ifstream ifs;
+    ifs.open(kFilename, std::ifstream::in);
+    if (!ifs.good())
+    {
+        CommitConfig();
+        ifs.open(kFilename, std::ifstream::in);
+    }
+    VerifyOrExit(ifs.is_open(), err = CHIP_ERROR_OPEN_FAILED);
```
Comment (andy31415) [inline examples/ota-requestor-app/linux/PersistentStorage.cpp:85]: usage of  return  macors would be prefferable: this does not  seem to have any exit cleanup code.

**Review by andy31415 — APPROVED** [review]
Approving based on SWTT developers working on OTA requests.

however comment says 'this app does not yet build' which is concerning - ideally we should check in code that at least compiles.

**Comment by andy31415** [comment]
NOTE: Approved since OTA development team (@carol-apple , @selissia , @kkasperczyk-no ?) seemed to want this in.

The part in the readme that says 'code does not yet compile' is worrysome and wondering if it would maybe be better to not check in code that does not compile and instead have work proceed on a separate branch. My checkmark is more of a 'I trust OTA team to do the right thing'.

File: examples/ota-requestor-app/linux/PersistentStorage.cpp (hunk 3 above)
Comment (holbrookt) [reply examples/ota-requestor-app/linux/PersistentStorage.cpp:85]: It can be removed completely once we are able to not use `Controller` code here

**Comment by holbrookt** [comment]
> NOTE: Approved since OTA development team (@carol-apple , @selissia , @kkasperczyk-no ?) seemed to want this in.
> 
> The part in the readme that says 'code does not yet compile' is worrysome and wondering if it would maybe be better to not check in code that does not compile and instead have work proceed on a separate branch. My checkmark is more of a 'I trust OTA team to do the right thing'.

I was asked by @woody-apple to submit whatever I currently had for the `ota-requestor-app` so that Carol and Sergei could get started helping out.

In any case, I am preparing a follow-up PR that will get this to build and run some minimal functionality ASAP.

**Review by chrisdecenzo — APPROVED** [review]

**Review by woody-apple — APPROVED** [review]

**Pull request merged**